lega/verify.py

```diff
@@ -19,102 +19,97 @@
-
-        # If client verification is required
-        certfile = CONF.get_value('quality_control', 'certfile', default=None)
-        if certfile:
-            LOG.debug("Prepare for client verification")
-            keyfile = CONF.get_value('quality_control', 'keyfile')
-            context.load_cert_chain(certfile, keyfile=keyfile)
-
-    try:
-        with urlopen(keyurl, context=context) as response:
-            assert(response.status == 200)
-            privkey = response.read()
-            if not privkey:  # Correcting a bug in the EGA keyserver
-                # When key not found, it returns a 200 and an empty payload.
-                # It should probably be changed to a 404
-                raise exceptions.PGPKeyError('No PGP key found')
-            return header_to_records(privkey, header, os.environ['LEGA_PASSWORD']), keyid
-    except HTTPError as e:
-        LOG.error(e)
-        msg = str(e)
-        if e.code == 404:  # If key not found, then probably wrong key.
-            raise exceptions.PGPKeyError(msg)
-        # Otherwise
-        raise exceptions.KeyserverError(msg)
-    # except Exception as e:
-    #     raise exceptions.KeyserverError(str(e))
+class ChecksumFile():
+    '''Fake IO writer, accepting bytes to checksum but not writing them anywhere'''
+    def __init__(self):
+        self.md = hashlib.sha256()
+
+    def write(self, data):
+        self.md.update(data)
+
+    def hexdigest(self):
+        return self.md.hexdigest()
+
+class PrependHeaderFile():
+    def __init__(self, header, bulk):
+        assert(header)
+        self.header = header
+        self.file = bulk
+        self.pos = 0
+        self.header_length = len(header)
+
+    def seek(self, offset, whence):
+        # Not needed because we decrypt all of it, and not only a range
+        raise NotImplementedError(f'Moving file pointer to {offset}: Unused case')
+
+    def read(self, size=-1):
+        # if size < 0: # read all
+        #     if self.pos >= self.length: # heade consumed already
+        #         return self.file.read()
+        #     return self.header[self.pos:] + self.file.read()
+        if size < 1:
+            raise NotImplementedError(f'Reading {size} bytes: Unused case')
+
+        if self.pos + size <= self.header_length:
+            res = self.header[self.pos:self.pos+size]
+            self.pos += size
+            return res
+
+        if self.pos + size > self.header_length:
+            if self.pos >= self.header_length: # already 
+                return self.file.read(size)
+
+            assert(self.header_length - self.pos > 0)
+            res = self.header[self.pos:]
+            self.pos += size
+            return res + self.file.read(size-(self.header_length - self.pos))
+
+    def readinto(self, b):
+        assert( isinstance(b, bytearray) )
+        data = self.read(len(b))
+        n = len(data)
+        b[:n] = data
+        return n
+
+    # def readinto(self, b):
```

Comment:
maybe miss ? similar function above

Reply:
The non-commented one is my version without involving memory views. It looks simpler and maybe more efficient. I kept the other one in case it was failing. Apparently it isn't.